### ISSUE TITLE
I.6, I.7 balance & improve these two rules

### ISSUE THREAD

**Issue by dksmiffs** [isocpp/CppCoreGuidelines#1384]
Suggest:
1. Improve I.6 **Example** with a reference to I.7 as follows:
```
int area(int height, int width)
{
    Expects(height > 0 && width > 0);            // good
    if (height <= 0 || width <= 0) my_error();   // obscure
    // ...
    // see I.7 for postcondition guidance
}
```
2. Improve I.7's first **Example, bad**'s "Consider using" code with a reference to I.6 as follows:
```
int area(int height, int width)
{
    // see I.6 for precondition guidance
    auto res = height * width;
    Ensures(res > 0);
    return res;
}
```
3. Copy I.7's introductory **Example, bad** to I.6, since it's relevant in both.  This would allow each rule to stand better on its own.
4. In I.7, move first explanatory sentence ("*Here, we (incautiously)...*") after **Example, bad** to I.6, since this sentence discusses preconditions, not postconditions.
5. In I.7 (and similarly in I.6, after implementing above suggestions), suggest rewording so that the improvement to the first **Example, bad** appears in a new **Example, better**, like this:
> ... Overflow can happen.
>
> **Example, better** Consider using:
>
> `int area(...`
6. In I.7's suggested improvement on the first example, `Ensures(res > 0);` is a check for addition overflow given 2 positive addends.  It is *not* a check for multiplication overflow (see http://www.cplusplus.com/articles/DE18T05o/).  This might have the unintended consequence of reducing the reader's confidence in the validity of this rule as a whole.
7. In general regarding these two rules, we know Contracts are on the horizon, but existing guidance on `Expects()` and `Ensures()` is *very* sparse via a simple Google search today.  These rules are the best place to clearly guide C++ developers in the recommended direction while we all await Contracts.

I believe suggestions in this issue would also bring improvement (but not closure) to issue #49, which also discusses I.7.

**Comment by MikeGitb** [comment]
Those rules really need an overhaul: Both examples in L.7 are wrong (the memset one dangerously so) and have been for years. I really don't understand, why the maintainers of this repository seem so callous about spreding misinformation. If you can't immediately think of a better example/formulation/guideline, then please, please at least remove the wrong stratements from the document and maybe leave a note.

**Comment by dksmiffs** [comment]
@MikeGitb , the memset issue with I.7 is already covered in #11.  I intentionally didn't repeat that issue here.

**Comment by hsutter** [comment]
Editors call: Agreed this connects again to #11. We have been delaying action on these because contracts have been progressing in standardization, and we feel those are the right tool. Now that they are adopted into draft C++20, and C++20 is within a year of shipping, we intend to revisit all the "expects and ensures" guidance as a whole in a contracts world.
